### PR TITLE
Eliminate allocation of dummy function  

### PR DESCRIPTION
**Pull request by hhugo**
The PR: 
- updates internalMod to not use the `.fun` property proxy for functions.
- adds an optimization to eliminate allocation of dummy function (`caml_alloc_dummy_function`). The dummy allocation is not necessary in JavaScript due to the weird non-lexical scoping.

Assuming the optimization properly eliminate all occurrences of `caml_alloc_dummy_function`,  we should be able to remove `.fun` property handling from the runtime.   

Note that `caml_alloc_dummy_function` was introduced in OCaml `4.02.3`, we would need to bump the lower bound from `4.02.0` to `4.02.3` which is fine.

### CODE REVIEW

File: compiler/tests-compiler/recfun.ml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+      (fun hsv ->
+         fun arg ->
+           match arg with
+           | Atom _a0 ->
+             let hsv = fold_int hsv 0 in
+             let hsv = hsv in hash_fold_string hsv _a0
+           | List _a0 ->
+             let hsv = fold_int hsv 1 in
+             let hsv = hsv in hash_fold_list hash_fold_t hsv _a0 : state -> t -> state)
+    and (hash : t -> hash_value) =
```
Comment (TyOverby) [inline compiler/tests-compiler/recfun.ml:91]: I'm not seeing this function in the output.  Did it get inlined?

File: runtime/internalMod.js [hunk 2]
```diff
@@ -53,22 +53,27 @@ function caml_CamlinternalMod_init_mod(loc,shape) {
...
+      switch(shape[0]){
+      case 0://module
+        for(var i=1;i<shape[1].length;i++)
+          loop(shape[1][i],real[i],x[i],real,i);
+        break;
+        //case 1://Value
+      default:
+      };
+  }
+  loop(shape,real,x,undefined,undefined);
```
Comment (TyOverby) [inline runtime/internalMod.js:77]: Why is there a loop now?

File: compiler/tests-compiler/recfun.ml (hunk 1 above)
Comment (hhugo) [reply compiler/tests-compiler/recfun.ml:91]: Yes

Comment (TyOverby) [reply compiler/tests-compiler/recfun.ml:91]: Could you add a test with functions where none of them get inlined?

File: runtime/internalMod.js (hunk 2 above)
Comment (hhugo) [reply runtime/internalMod.js:77]: Update_mod was already recursive. We introduce this inner recursive function loop to keep track of parent and i and allow to  more direct update for function

File: compiler/tests-compiler/recfun.ml (hunk 1 above)
Comment (hhugo) [reply compiler/tests-compiler/recfun.ml:91]: Note that the inlining was enabled by the last commit. Probably better the review per commits

Comment (hhugo) [reply compiler/tests-compiler/recfun.ml:91]: I've tune the test to prevent inlining

**Comment by hhugo** [comment]
@xclerc, do you want to take a look at this one ?

**Comment by xclerc** [comment]
Just to be sure I understand the optimization: the idea is that every time we have
both `x = caml_alloc_dummy_function(...)` and `caml_update_dummy(x, y)`, we
can get rid of both function calls and use `y` instead of `x`, right?

It looks like [this call to filter](https://github.com/ocsigen/js_of_ocaml/pull/1013/files#diff-d36f6c73c98b086e0c53e7ce796c8703R110) is expected to be the identity. My understanding is
that it might be true only given the current state of the code generator of `ocamlc`
(`Bytegen.comp_expr`), but then if we want to be independent of the code generator,
wouldn't it be safer to detect e.g. several `caml_update_dummy(x, y_i)` calls?

**Comment by hhugo** [comment]
> Just to be sure I understand the optimization: the idea is that every time we have
both x = caml_alloc_dummy_function(...) and caml_update_dummy(x, y), we
can get rid of both function calls and use y instead of x, right?

Correct

> It looks like this call to filter is expected to be the identity. My understanding is
that it might be true only given the current state of the code generator of ocamlc
(Bytegen.comp_expr), but then if we want to be independent of the code generator,
wouldn't it be safer to detect e.g. several caml_update_dummy(x, y_i) calls?

That sound correct. We should at least add some assertion.

**Comment by hhugo** [comment]
For the record, caml_alloc_dummy_function was introduced in 4.02.3

**Comment by hhugo** [comment]
> It looks like this call to filter is expected to be the identity.

Looking more closely at the code, it's not the case. [filter] is here to only keep info from [caml_update_dummy] that have a corresponding [caml_alloc_dummy_function]. There are other cases (e.g. `let rec l = 0 :: l`).

I've updated the code with comments.
